src/cdomains/floatDomain.ml

```diff
@@ -37,130 +38,210 @@ module type FloatDomainBase = sig
 
-  let arbitrary () = failwith "todo"
+  (**for QCheck: should describe how to generate random values and shrink possilbe counter examples *)
+  let arbitrary () = QCheck.map norm_arb (QCheck.option (QCheck.pair QCheck.float QCheck.float)) 
 
-  let relift (x : t) = failwith "todo"
+  let relift x = x 
 
-  let leq _ _ = failwith "todo"
+  let leq v1 v2 = 
+    match v1, v2 with
+    | _, None -> true
+    | None, Some _ -> false
+    | Some (l1, h1), Some (l2, h2) -> l1 >= l2 && h1 <= h2
 
-  let join _ _ = failwith "todo"
+  let join v1 v2 = 
+    match v1, v2 with
+    | None, v | v, None -> None
+    | Some (l1, h1), Some (l2, h2) -> Some (min l1 l2, max h1 h2)
 
-  let meet _ _ = failwith "todo"
+  let meet v1 v2 = 
+    match v1, v2 with
+    | None, v | v, None -> v
+    | Some (l1, h1), Some (l2, h2) -> 
+      let (l, h) = (max l1 l2, min h1 h2) in
+      if l <= h 
+      then Some (l, h) 
+      else failwith "meet results in empty Interval"
 
   (** [widen x y] assumes [leq x y]. Solvers guarantee this by calling [widen old (join old new)]. *)
-  let widen _ _ = failwith "todo"
-
-  let narrow _ _ = failwith "todo"
+  let widen v1 v2 = (**TODO: support 'threshold_widening' option *)
+    match v1, v2 with
+    | None, _ | _, None -> None
+    | Some (l1, h1), Some (l2, h2) ->
+      (**If we widen and we know that neither interval contains +-inf or nan, it is ok to widen only to +-max_float, 
+         because a widening with +-inf/nan will always result in the case above -> Top *)
+      let low = if l1 <= l2 then l1 else (-.Float.max_float) in 
+      let high = if h1 >= h2 then h1 else Float.max_float in
+      norm @@ Some (low, high)
+
+  let narrow v1 v2 = (**TODO: support 'threshold_narrowing' and 'narrow_by_meet' option *)
+    match v1, v2 with (**we cannot distinguish between the lower bound beeing -inf or the upper bound beeing inf. Also there is nan *)
+    | None, _ -> v2
+    | _, _ -> v1
 
   (** If [leq x y = false], then [pretty_diff () (x, y)] should explain why. *)
   let pretty_diff () (x, y) =
     Pretty.dprintf "%a instead of %a" pretty x pretty y
 
   let bot () = failwith "no bot exists"
 
-  let is_bot _ = failwith "no bot exists"
+  let is_bot _ = false
 
   let top () = None
 
   let is_top = Option.is_none
 
   let neg = Option.map (fun (low, high) -> (-.high, -.low))
 
-  let add = eval_binop add
-
-  let sub = eval_binop sub
-
-  let mul = eval_binop mul
-
-  let div = eval_binop div
-
-  let lt = eval_int_binop lt
-
-  let gt = eval_int_binop gt
-
-  let le _ _ = failwith "todo - le"
-
-  let ge _ _ = failwith "todo - ge"
-
-  let eq = eval_int_binop eq
-
-  let ne = eval_int_binop ne
+  let add op1 op2 = 
+    match op1, op2 with
+    | Some (l1, h1), Some (l2, h2) -> Some (add Down l1 l2, add Up h1 h2)
+    | _ -> None
+
+  let sub op1 op2 = 
+    match op1, op2 with
```

Comment:
I first removed the helper method, because I thought it was a bit confusing. However, I see the benefit and re-added it, but moved it into the FloatInterval module to make clearer where it belongs